src/utils.rs

```diff
@@ -177,6 +177,81 @@ pub fn canonicalize(path: impl AsRef<Path>) -> Result<PathBuf, SolcIoError> {
+///
+/// See also: <https://docs.soliditylang.org/en/v0.8.23/path-resolution.html>
+pub fn normalize_solidity_import_path(
+    directory: impl AsRef<Path>,
+    import_path: impl AsRef<Path>,
+) -> Result<PathBuf, SolcIoError> {
+    let original = directory.as_ref().join(import_path);
+    let cleaned = clean_solidity_path(&original);
+
+    // this is to align the behavior with `canonicalize`
+    use path_slash::PathExt;
+    let normalized = PathBuf::from(dunce::simplified(&cleaned).to_slash_lossy().as_ref());
+
+    fs::metadata(&normalized).map(|_| normalized).map_err(|err| SolcIoError::new(err, original))
+}
+
+// This function lexically cleans the given path.
+//
+// It performs the following transformations for the path:
+//
+// * Resolves references (current directories (`.`) and parent (`..`) directories).
+// * Reduces repeated separators to a single separator (e.g., from `//` to `/`).
+//
+// This transformation is lexical, not involving the file system, which means it does not account
+// for symlinks. This approach has a caveat. For example, consider a filesystem-accessible path
+// `a/b/../c.sol` passed to this function. It returns `a/c.sol`. However, if `b` is a symlink,
+// `a/c.sol` might not be accessible in the filesystem in some environments. Despite this, it's
+// unlikely that this will pose a problem for our intended use.
+//
+// # How it works
+//
+// The function splits the given path into components, where each component roughly corresponds to a
+// string between separators. It then iterates over these components (starting from the leftmost
+// part of the path) to reconstruct the path. The following steps are applied to each component:
+//
+// * If the component is a current directory, it's removed.
+// * If the component is a parent directory, the following rules are applied:
+//     * If the preceding component is a normal, then both the preceding normal component and the
+//       parent directory component are removed. (Examples of normal components include `a` and `b`
+//       in `a/b`.)
+//     * Otherwise (if there is no preceding component, or if the preceding component is a parent,
+//       root, or prefix), it remains untouched.
+// * Otherwise, the component remains untouched.
+//
+// Finally, the processed components are reassembled into a path.
+fn clean_solidity_path(original_path: impl AsRef<Path>) -> PathBuf {
+    let mut new_path = Vec::new();
+
+    for component in original_path.as_ref().components() {
+        match component {
+            Component::Prefix(..) | Component::RootDir | Component::Normal(..) => {
+                new_path.push(component);
+            }
+            Component::CurDir => {}
+            Component::ParentDir => match new_path.last() {
+                Some(Component::Normal(..)) => {
+                    new_path.pop();
+                }
+                _ => {
+                    new_path.push(component);
+                }
```

Comment:
I think we can use `if let Some(Component::Normal) = new_path.last() {pop} else {...}` here

which is a bit nicer imo